File: android/src/main/java/cn/rongcloud/rongcloud_rtc_wrapper_plugin/ArgumentAdapter.java
package cn.rongcloud.rongcloud_rtc_wrapper_plugin;

import androidx.annotation.NonNull;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import cn.rongcloud.rtc.wrapper.config.RCRTCIWAudioConfig;
import cn.rongcloud.rtc.wrapper.config.RCRTCIWCustomLayout;
import cn.rongcloud.rtc.wrapper.config.RCRTCIWVideoConfig;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWAudioCodecType;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWAudioMixingMode;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWAudioQuality;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWAudioScenario;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWCamera;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWCameraCaptureOrientation;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWLiveMixLayoutMode;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWLiveMixRenderMode;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWLocalAudioStats;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWLocalVideoStats;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWMediaType;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWNetworkStats;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWRemoteAudioStats;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWRemoteVideoStats;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWRole;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWVideoFps;
import cn.rongcloud.rtc.wrapper.constants.RCRTCIWVideoResolution;
import cn.rongcloud.rtc.wrapper.setup.RCRTCIWAudioSetup;
import cn.rongcloud.rtc.wrapper.setup.RCRTCIWEngineSetup;
import cn.rongcloud.rtc.wrapper.setup.RCRTCIWRoomSetup;
import cn.rongcloud.rtc.wrapper.setup.RCRTCIWVideoSetup;

/**
 * @author panmingda
 * @date 2021/6/9
 */
final class ArgumentAdapter {

    private ArgumentAdapter() {
    }

    static RCRTCIWMediaType toMediaType(@NonNull Integer type) {
        return RCRTCIWMediaType.values()[type];
    }

    static RCRTCIWCamera toCamera(@NonNull Integer camera) {
        return RCRTCIWCamera.values()[camera];
    }

    static RCRTCIWCameraCaptureOrientation toCameraCaptureOrientation(@NonNull Integer orientation) {
        return RCRTCIWCameraCaptureOrientation.values()[orientation];
    }

    static RCRTCIWLiveMixLayoutMode toLiveMixLayoutMode(@NonNull Integer mode) {
        return RCRTCIWLiveMixLayoutMode.values()[mode];
    }

    static RCRTCIWLiveMixRenderMode toLiveMixRenderMode(@NonNull Integer mode) {
        return RCRTCIWLiveMixRenderMode.values()[mode];
    }

    static RCRTCIWVideoResolution toVideoResolution(@NonNull Integer resolution) {
        return RCRTCIWVideoResolution.values()[resolution];
    }

    static RCRTCIWVideoFps toVideoFps(@NonNull Integer fps) {
        return RCRTCIWVideoFps.values()[fps];
    }

    static RCRTCIWAudioMixingMode toAudioMixingMode(@NonNull Integer mode) {
        return RCRTCIWAudioMixingMode.values()[mode];
    }

    static RCRTCIWAudioCodecType toAudioCodecType(@NonNull Integer type) {
        return RCRTCIWAudioCodecType.values()[type];
    }

    static RCRTCIWRole toRole(@NonNull Integer role) {
        return RCRTCIWRole.values()[role];
    }

    static RCRTCIWAudioQuality toAudioQuality(@NonNull Integer quality) {
        return RCRTCIWAudioQuality.values()[quality];
    }

    static RCRTCIWAudioScenario toAudioScenario(@NonNull Integer scenario) {
        return RCRTCIWAudioScenario.values()[scenario];
    }

    static RCRTCIWAudioSetup toAudioSetup(@NonNull HashMap<String, Object> setup) {
        Integer codec = (Integer) setup.get("codec");
        assert (codec != null);
        Integer audioSource = (Integer) setup.get("audioSource");
        assert (audioSource != null);
        Integer audioSampleRate = (Integer) setup.get("audioSampleRate");
        assert (audioSampleRate != null);
        Boolean enableMicrophone = (Boolean) setup.get("enableMicrophone");
        assert (enableMicrophone != null);
        Boolean enableStereo = (Boolean) setup.get("enableStereo");
        assert (enableStereo != null);
        return RCRTCIWAudioSetup.Builder.create()
                .withAudioCodecType(toAudioCodecType(codec))
                .withAudioSource(audioSource)
                .withAudioSampleRate(audioSampleRate)
                .withEnableMicrophone(enableMicrophone)
                .withEnableStereo(enableStereo)
                .build();
    }

    static RCRTCIWVideoSetup toVideoSetup(@NonNull HashMap<String, Object> setup) {
        Boolean enableHardwareDecoder = (Boolean) setup.get("enableHardwareDecoder");
        assert (enableHardwareDecoder != null);
        Boolean enableHardwareEncoder = (Boolean) setup.get("enableHardwareEncoder");
        assert (enableHardwareEncoder != null);
        Boolean enableHardwareEncoderHighProfile = (Boolean) setup.get("enableHardwareEncoderHighProfile");
        assert (enableHardwareEncoderHighProfile != null);
        Integer hardwareEncoderFrameRate = (Integer) setup.get("hardwareEncoderFrameRate");
        assert (hardwareEncoderFrameRate != null);
        Boolean enableEncoderTexture = (Boolean) setup.get("enableEncoderTexture");
        assert (enableEncoderTexture != null);
        Boolean enableTinyStream = (Boolean) setup.get("enableTinyStream");
        assert (enableTinyStream != null);
        return RCRTCIWVideoSetup.Builder.create()
                .withEnableHardwareDecoder(enableHardwareDecoder)
                .withEnableHardwareEncoder(enableHardwareEncoder)
                .withEnableHardwareEncoderHighProfile(enableHardwareEncoderHighProfile)
                .withHardwareEncoderFrameRate(hardwareEncoderFrameRate)
                .withEnableEncoderTexture(enableEncoderTexture)
                .withEnableTinyStream(enableTinyStream)
                .build();
    }

    static RCRTCIWEngineSetup toEngineSetup(@NonNull HashMap<String, Object> setup) {
        Boolean reconnectable = (Boolean) setup.get("reconnectable");
        assert (reconnectable != null);
        Integer statsReportInterval = (Integer) setup.get("statsReportInterval");
        assert (statsReportInterval != null);
        RCRTCIWEngineSetup.Builder builder = RCRTCIWEngineSetup.Builder.create()
                .withReconnectable(reconnectable)
                .withStatusReportInterval(statsReportInterval);
        Object mediaUrl = setup.get("mediaUrl");
        if (mediaUrl != null) {
            builder.withMediaUrl((String) mediaUrl);
        }
        Object audioSetup = setup.get("audioSetup");
        if (audioSetup != null) {
            builder.withAudioSetup(toAudioSetup((HashMap<String, Object>) audioSetup));
        }
        Object videoSetup = setup.get("videoSetup");
        if (videoSetup != null) {
            builder.withVideoSetup(toVideoSetup((HashMap<String, Object>) videoSetup));
        }
        return builder.build();
    }

    @SuppressWarnings("unchecked")
    static RCRTCIWRoomSetup toRoomSetup(@NonNull HashMap<String, Object> setup) {
        Integer type = (Integer) setup.get("type");
        assert (type != null);
        Integer role = (Integer) setup.get("role");
        assert (role != null);
        RCRTCIWRoomSetup.Builder builder = RCRTCIWRoomSetup.Builder.create()
                .withMediaType(toMediaType(type))
                .withRole(toRole(role));
        return builder.build();
    }

    static RCRTCIWAudioConfig toAudioConfig(@NonNull HashMap<String, Object> config) {
        Integer quality = (Integer) config.get("quality");
        assert (quality != null);
        Integer scenario = (Integer) config.get("scenario");
        assert (scenario != null);
        return RCRTCIWAudioConfig.create()
                .setQuality(toAudioQuality(quality))
                .setScenario(toAudioScenario(scenario));
    }

    static RCRTCIWVideoConfig toVideoConfig(@NonNull HashMap<String, Object> config) {
        Integer minBitrate = (Integer) config.get("minBitrate");
        assert (minBitrate != null);
        Integer maxBitrate = (Integer) config.get("maxBitrate");
        assert (maxBitrate != null);
        Integer fps = (Integer) config.get("fps");
        assert (fps != null);
        Integer resolution = (Integer) config.get("resolution");
        assert (resolution != null);
        Boolean mirror = (Boolean) config.get("mirror");
        assert (mirror != null);
        return RCRTCIWVideoConfig.create()
                .setMinBitrate(minBitrate)
                .setMaxBitrate(maxBitrate)
                .setFps(toVideoFps(fps))
                .setResolution(toVideoResolution(resolution))
                .setMirror(mirror);
    }

    static List<RCRTCIWCustomLayout> toLiveMixCustomLayouts(@NonNull List<HashMap<String, Object>> layouts) {
        List<RCRTCIWCustomLayout> lists = new ArrayList<>(layouts.size());
        for (HashMap<String, Object> layout : layouts) {
            lists.add(toLiveMixCustomLayout(layout));
        }
        return lists;
    }

    static RCRTCIWCustomLayout toLiveMixCustomLayout(@NonNull HashMap<String, Object> layout) {
        String id = (String) layout.get("id");
        assert (id != null);
        Integer x = (Integer) layout.get("x");
        assert (x != null);
        Integer y = (Integer) layout.get("y");
        assert (y != null);
        Integer width = (Integer) layout.get("width");
        assert (width != null);
        Integer height = (Integer) layout.get("height");
        assert (height != null);
        return new RCRTCIWCustomLayout(id, x, y, width, height);
    }

    static HashMap<String, Object> fromNetworkStats(@NonNull RCRTCIWNetworkStats stats) {
        HashMap<String, Object> map = new HashMap<>();
        map.put("type", stats.getType().ordinal());
        map.put("ip", stats.getIp());
        map.put("sendBitrate", stats.getSendBitrate());
        map.put("receiveBitrate", stats.getReceiveBitrate());
        map.put("rtt", stats.getRtt());
        return map;
    }

    static HashMap<String, Object> fromLocalAudioStats(@NonNull RCRTCIWLocalAudioStats stats) {
        HashMap<String, Object> map = new HashMap<>();
        map.put("codec", stats.getCodec().ordinal());
        map.put("bitrate", stats.getBitrate());
        map.put("volume", stats.getVolume());
        map.put("packageLostRate", stats.getPackageLostRate());
        map.put("rtt", stats.getRtt());
        return map;
    }

    static HashMap<String, Object> fromLocalVideoStats(@NonNull RCRTCIWLocalVideoStats stats) {
        HashMap<String, Object> map = new HashMap<>();
        map.put("tiny", stats.isTiny());
        map.put("codec", stats.getCodec().ordinal());
        map.put("bitrate", stats.getBitrate());
        map.put("fps", stats.getFps());
        map.put("width", stats.getWidth());
        map.put("height", stats.getHeight());
        map.put("packageLostRate", stats.getPackageLostRate());
        map.put("rtt", stats.getRtt());
        return map;
    }

    static HashMap<String, Object> fromRemoteAudioStats(@NonNull RCRTCIWRemoteAudioStats stats) {
        HashMap<String, Object> map = new HashMap<>();
        map.put("id", stats.getUserId());
        map.put("codec", stats.getCodec().ordinal());
        map.put("bitrate", stats.getBitrate());
        map.put("volume", stats.getVolume());
        map.put("packageLostRate", stats.getPackageLostRate());
        map.put("rtt", stats.getRtt());
        return map;
    }

    static HashMap<String, Object> fromRemoteVideoStats(@NonNull RCRTCIWRemoteVideoStats stats) {
        HashMap<String, Object> map = new HashMap<>();
        map.put("id", stats.getUserId());
        map.put("codec", stats.getCodec().ordinal());
        map.put("bitrate", stats.getBitrate());
        map.put("fps", stats.getFps());
        map.put("width", stats.getWidth());
        map.put("height", stats.getHeight());
        map.put("packageLostRate", stats.getPackageLostRate());
        map.put("rtt", stats.getRtt());
        return map;
    }
}

File: ios/Classes/ArgumentAdapter.m
//
//  ArgumentAdapter.m
//  rongcloud_rtc_wrapper_plugin
//
//  Created by 潘铭达 on 2021/6/11.
//

#import "ArgumentAdapter.h"

RCRTCIWMediaType toMediaType(int type) {
    return (RCRTCIWMediaType) type;
}

RCRTCIWCamera toCamera(int camera) {
    return (RCRTCIWCamera) (camera - 1);
}

RCRTCIWCameraCaptureOrientation toCameraCaptureOrientation(int orientation) {
    return (RCRTCIWCameraCaptureOrientation) (orientation + 1);
}

RCRTCIWLiveMixLayoutMode toLiveMixLayoutMode(int mode) {
    return (RCRTCIWLiveMixLayoutMode) (mode + 1);
}

RCRTCIWLiveMixRenderMode toLiveMixRenderMode(int mode) {
    return (RCRTCIWLiveMixRenderMode) (mode + 1);
}

RCRTCIWVideoResolution toVideoResolution(int resolution) {
    return (RCRTCIWVideoResolution) resolution;
}

RCRTCIWVideoFps toVideoFps(int fps) {
    return (RCRTCIWVideoFps) fps;
}

RCRTCIWAudioMixingMode toAudioMixingMode(int mode) {
    return (RCRTCIWAudioMixingMode) mode;
}

RCRTCIWAudioCodecType toAudioCodecType(int type) {
    switch (type) {
        case 1:
            return RCRTCIWAudioCodecTypeOPUS;
        default:
            return RCRTCIWAudioCodecTypePCMU;
    }
}

RCRTCIWRole toRole(int role) {
    return (RCRTCIWRole) role;
}

RCRTCIWAudioQuality toAudioQuality(int quality) {
    switch (quality) {
        case 1:
            return RCRTCIWAudioQualityMusic;
        case 2:
            return RCRTCIWAudioQualityMusicHigh;
        default:
            return RCRTCIWAudioQualitySpeech;
    }
}

RCRTCIWAudioScenario toAudioScenario(int scenario) {
    return (RCRTCIWAudioScenario) scenario;
}

RCRTCIWAudioSetup *toAudioSetup(NSDictionary *arguments) {
    int codec = [arguments[@"codec"] intValue];
    RCRTCIWAudioSetup *setup = [[RCRTCIWAudioSetup alloc] init];
    setup.type = toAudioCodecType(codec);
    return setup;
}

RCRTCIWVideoSetup *toVideoSetup(NSDictionary *arguments) {
    bool enableTinyStream = [arguments[@"enableTinyStream"] boolValue];
    RCRTCIWVideoSetup *setup = [[RCRTCIWVideoSetup alloc] init];
    setup.enableTinyStream = enableTinyStream;
    return setup;
}

RCRTCIWEngineSetup *toEngineSetup(NSDictionary *arguments) {
    bool reconnectable = [arguments[@"reconnectable"] boolValue];
    int statsReportInterval = [arguments[@"statsReportInterval"] intValue];
    id mediaUrl = arguments[@"mediaUrl"];
    id audioSetup = arguments[@"audioSetup"];
    id videoSetup = arguments[@"videoSetup"];
    RCRTCIWEngineSetup *setup = [[RCRTCIWEngineSetup alloc] init];
    setup.reconnectable = reconnectable;
    setup.statsReportInterval = statsReportInterval;
    if (![mediaUrl isEqual:[NSNull null]]) {
        setup.mediaUrl = mediaUrl;
    }
    if (![audioSetup isEqual:[NSNull null]]) {
        setup.audioSetup = toAudioSetup(audioSetup);
    }
    if (![videoSetup isEqual:[NSNull null]]) {
        setup.videoSetup = toVideoSetup(videoSetup);
    }
    return setup;
}

RCRTCIWRoomSetup *toRoomSetup(NSDictionary *arguments) {
    int type = [arguments[@"type"] intValue];
    int role = [arguments[@"role"] intValue];
    RCRTCIWRoomSetup *setup = [[RCRTCIWRoomSetup alloc] init];
    setup.type = toMediaType(type);
    setup.role = toRole(role);
    return setup;
}

RCRTCIWAudioConfig *toAudioConfig(NSDictionary *arguments) {
    int quality = [arguments[@"quality"] intValue];
    int scenario = [arguments[@"scenario"] intValue];
    RCRTCIWAudioConfig *config = [[RCRTCIWAudioConfig alloc] init];
    config.quality = toAudioQuality(quality);
    config.scenario = toAudioScenario(scenario);
    return config;
}

RCRTCIWVideoConfig *toVideoConfig(NSDictionary *arguments) {
    int minBitrate = [arguments[@"minBitrate"] intValue];
    int maxBitrate = [arguments[@"maxBitrate"] intValue];
    int fps = [arguments[@"fps"] intValue];
    int resolution = [arguments[@"resolution"] intValue];
    bool mirror = [arguments[@"mirror"] boolValue];
    RCRTCIWVideoConfig *config = [[RCRTCIWVideoConfig alloc] init];
    config.minBitrate = minBitrate;
    config.maxBitrate = maxBitrate;
    config.fps = toVideoFps(fps);
    config.resolution = toVideoResolution(resolution);
    config.mirror = mirror;
    return config;
}

RCRTCIWCustomLayout *toLiveMixCustomLayout(NSDictionary *arguments) {
    NSString *uid = arguments[@"id"];
    int x = [arguments[@"x"] intValue];
    int y = [arguments[@"y"] intValue];
    int width = [arguments[@"width"] intValue];
    int height = [arguments[@"height"] intValue];
    RCRTCIWCustomLayout *layout = [[RCRTCIWCustomLayout alloc] init];
    layout.userId = uid;
    layout.x = x;
    layout.y = y;
    layout.width = width;
    layout.height = height;
    return layout;
}

NSArray<RCRTCIWCustomLayout *> *toLiveMixCustomLayouts(NSArray<NSDictionary *> *arguments) {
    NSMutableArray *layouts = [NSMutableArray array];
    for (NSDictionary *argument in arguments) {
        [layouts addObject:toLiveMixCustomLayout(argument)];
    }
    return layouts;
}

NSDictionary *fromNetworkStats(RCRTCIWNetworkStats *stats) {
    NSMutableDictionary *dictionary = [NSMutableDictionary dictionary];
    [dictionary setObject:@((int) stats.type) forKey:@"type"];
    [dictionary setObject:stats.ip forKey:@"ip"];
    [dictionary setObject:@(stats.sendBitrate) forKey:@"sendBitrate"];
    [dictionary setObject:@(stats.receiveBitrate) forKey:@"receiveBitrate"];
    [dictionary setObject:@(stats.rtt) forKey:@"rtt"];
    return dictionary;
}

int fromAudioCodecType(RCRTCIWAudioCodecType type) {
    switch (type) {
        case RCRTCIWAudioCodecTypePCMU:
            return 0;
        default:
            return 1;
    }
}

NSDictionary *fromLocalAudioStats(RCRTCIWLocalAudioStats *stats) {
    NSMutableDictionary *dictionary = [NSMutableDictionary dictionary];
    [dictionary setObject:@(fromAudioCodecType(stats.codec)) forKey:@"codec"];
    [dictionary setObject:@(stats.bitrate) forKey:@"bitrate"];
    [dictionary setObject:@(stats.volume) forKey:@"volume"];
    [dictionary setObject:@(stats.packageLostRate) forKey:@"packageLostRate"];
    [dictionary setObject:@(stats.rtt) forKey:@"rtt"];
    return dictionary;
}

NSDictionary *fromLocalVideoStats(RCRTCIWLocalVideoStats *stats) {
    NSMutableDictionary *dictionary = [NSMutableDictionary dictionary];
    [dictionary setObject:@(stats.tiny) forKey:@"tiny"];
    [dictionary setObject:@((int) stats.codec) forKey:@"codec"];
    [dictionary setObject:@(stats.bitrate) forKey:@"bitrate"];
    [dictionary setObject:@(stats.fps) forKey:@"fps"];
    [dictionary setObject:@(stats.width) forKey:@"width"];
    [dictionary setObject:@(stats.height) forKey:@"height"];
    [dictionary setObject:@(stats.packageLostRate) forKey:@"packageLostRate"];
    [dictionary setObject:@(stats.rtt) forKey:@"rtt"];
    return dictionary;
}

NSDictionary *fromRemoteAudioStats(RCRTCIWRemoteAudioStats *stats) {
    NSMutableDictionary *dictionary = [NSMutableDictionary dictionary];
    [dictionary setObject:stats.userId forKey:@"id"];
    [dictionary setObject:@(fromAudioCodecType(stats.codec)) forKey:@"codec"];
    [dictionary setObject:@(stats.bitrate) forKey:@"bitrate"];
    [dictionary setObject:@(stats.volume) forKey:@"volume"];
    [dictionary setObject:@(stats.packageLostRate) forKey:@"packageLostRate"];
    [dictionary setObject:@(stats.rtt) forKey:@"rtt"];
    return dictionary;
}

NSDictionary *fromRemoteVideoStats(RCRTCIWRemoteVideoStats *stats) {
    NSMutableDictionary *dictionary = [NSMutableDictionary dictionary];
    [dictionary setObject:stats.userId forKey:@"id"];
    [dictionary setObject:@((int) stats.codec) forKey:@"codec"];
    [dictionary setObject:@(stats.bitrate) forKey:@"bitrate"];
    [dictionary setObject:@(stats.fps) forKey:@"fps"];
    [dictionary setObject:@(stats.width) forKey:@"width"];
    [dictionary setObject:@(stats.height) forKey:@"height"];
    [dictionary setObject:@(stats.packageLostRate) forKey:@"packageLostRate"];
    [dictionary setObject:@(stats.rtt) forKey:@"rtt"];
    return dictionary;
}
